[Telemetry-Otel] Added support for OtlpGrpcSpanExporter exporter

Shephalimittal opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
 Added support for OtlpGrpcSpanExporter. OtelSpanExporterFactory expects exporter has method create() (For details [link](https://github.com/opensearch-project/OpenSearch/blob/main/plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OTelSpanExporterFactory.java#L60)) OtlpGrpcSpanExporterProvider creates an instance of OtlpGrpcSpanExporter with https endpoint. 

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Shephalimittal commented:
@reta @shwetathareja @Gaganjuneja Please review.

Gaganjuneja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OtlpGrpcSpanExporterProvider.java:
```diff
@@ -0,0 +1,52 @@
+     * OtlpGrpcSpanExporter and sets https endpoint.
+     */
+    public static OtlpGrpcSpanExporterProvider create() {
+        OtlpGrpcSpanExporter exporter = OtlpGrpcSpanExporter.builder().setEndpoint("https://localhost:4317").build();
```
It should be configurable through OTelTelemetrySettings.

Gaganjuneja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OtlpGrpcSpanExporterProvider.java:
```diff
@@ -0,0 +1,52 @@
+/**
+* OtlpGrpcSpanExporterProvider class creates an instance of OtlpGrpcSpanExporter
+*/
+public class OtlpGrpcSpanExporterProvider implements SpanExporter {
```
Final class.

Gaganjuneja commented:
Add test cases in OTelSpanExporterFactoryTests and for the provider if needed.

reta commented:
> @reta @shwetathareja @Gaganjuneja Please review.

@Shephalimittal apologies, I do not have time left this week to review it, but will pick it up next week first thing

Gaganjuneja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OTelSpanExporterFactory.java:
```diff
@@ -45,21 +45,41 @@ private OTelSpanExporterFactory() {
-                        .asType(MethodType.methodType(SpanExporter.class))
-                        .invokeExact();
+                    String methodName = "create";
+                    boolean createMethodWithSettingsParamExists = true;
```
Please follow boolean variable naming convention e.g. isCreateMethodContainsSettingsParam

Gaganjuneja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OTelSpanExporterFactory.java:
```diff
@@ -45,21 +45,41 @@ private OTelSpanExporterFactory() {
+                        // Attempt to get the method with the specified name and parameter types
+                        spanExporterProviderClass.getMethod(methodName, Settings.class);
+                    } catch (NoSuchMethodException e) {
+                        createMethodWithSettingsParamExists = false;
```
Better check the parameter type and don't decide on the exception. Sample code here - https://github.com/opensearch-project/OpenSearch/blob/d375e4cfea1a8e4c58dd1409365a67cf49366780/modules/lang-painless/src/main/java/org/opensearch/painless/ScriptClassInfo.java#L132

shwetathareja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/OTelTelemetryPlugin.java:
```diff
@@ -44,7 +44,8 @@ public List<Setting<?>> getSettings() {
             OTelTelemetrySettings.TRACER_EXPORTER_MAX_QUEUE_SIZE_SETTING,
-            OTelTelemetrySettings.OTEL_TRACER_SPAN_EXPORTER_CLASS_SETTING
+            OTelTelemetrySettings.OTEL_TRACER_SPAN_EXPORTER_CLASS_SETTING,
+            OTelTelemetrySettings.TRACER_SPAN_EXPORTER_ENDPOINT
```
how is the protocol specified as gRPC?

shwetathareja replied:
Also OTel allows passing any specific headers, how would those be passed here?

Shephalimittal replied:
The Provider is just a wrapper over existing [OtlpGrpcSpanExporter](https://github.com/open-telemetry/opentelemetry-java/blob/2e2c308293800901aed771c3e56bba71b419257a/exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/trace/OtlpGrpcSpanExporter.java#L19). This uses gRPC. 

For passing specific settings, we need to make telemetrysdk as Auto-configurable. We will take that in later PRs https://github.com/opensearch-project/OpenSearch/pull/9453 after opening RFC.

Shephalimittal replied:
We will soon open RFC on how to make it auto configurable. For now, as part of experimental release we just want to give a way to visualise the spans so for now its fixed.

shwetathareja replied:
Is the change pushed to mark Request tracing as experimental via feature flag?

Shephalimittal replied:
yes

shwetathareja replied:
Please create a github issue to track this.

Shephalimittal replied:
https://github.com/opensearch-project/OpenSearch/issues/9764

shwetathareja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OTelSpanExporterFactory.java:
```diff
@@ -45,21 +46,40 @@ private OTelSpanExporterFactory() {
             SpecialPermission.check();
             return AccessController.doPrivileged((PrivilegedExceptionAction<SpanExporter>) () -> {
+                String methodName = "create";
+                boolean isCreateMethodContainsSettingsParam = false;
```
[minor]  createMethodContainsSettingsParam

shwetathareja commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OtlpGrpcSpanExporterProvider.java:
```diff
@@ -0,0 +1,64 @@
+        String endpoint = OTelTelemetrySettings.TRACER_SPAN_EXPORTER_ENDPOINT.get(settings);
+        // if endpoint is empty, do not set Endpoint
+        if (endpoint.isEmpty()) {
+            exporter = OtlpGrpcSpanExporter.builder().build();
```
what is the default endpoint used when it is not specified explicitly? Also what happens if the endpoint is down? would the traces be discarded?

Shephalimittal replied:
The default endpoint is "http://localhost:4317" [link](https://github.com/open-telemetry/opentelemetry-java/blob/2e2c308293800901aed771c3e56bba71b419257a/exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/trace/OtlpGrpcSpanExporterBuilder.java#L32).

If endpoint is down we will get these exceptions in log 
Failed to export spans. The request could not be executed. Full error message:

yes traces would be discarded,

shwetathareja replied:
Please call the default endpoint explicitly.

shwetathareja replied:
These spans are exported in async fashion and how would it impact the OpenSearch process in case exporter is running slow.

shwetathareja commented:
Waiting for gradle check to fix, before merging the PR

reta commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OtlpGrpcSpanExporterProvider.java:
```diff
@@ -0,0 +1,64 @@
+/**
+* OtlpGrpcSpanExporterProvider class creates an instance of OtlpGrpcSpanExporter
+*/
+public final class OtlpGrpcSpanExporterProvider implements SpanExporter {
```
@Shephalimittal @shwetathareja this is not the extensible way to deal with OTEL span exporters: we basically need to use wrappers for each and every one of them. I understand that we would like to have some initial implementation that work but we need also to think a bit into the future so we won't be asking people to go over painful settings migrations.

I would susggest:
  - use OtlpGrpcSpanExporter directly 
  - use (for now) `telemetry.otel.tracer.exporter.otlphrpcspanexporter.endpoint` setting (instead of generic one `telemetry.otel.tracer.exporter.endpoint`, not all exporters deal with that)
  - `OTelSpanExporterFactory` should recognize `<SpanExporterCass>::builder` class 
  - `OTelSpanExporterFactory` should extract   `telemetry.otel.tracer.exporter.<SpanExporterCass>.*` settings and call respective builder setters

This is not 100% bullet proof but would later prevent at least setting changes.

reta replied:
The `telemetry.otel.tracer.exporter.otlpgrpcspanexporter.endpoint` is specific to `OtlpGrpcSpanExporter` and no other exporters. Jaeger or Zipkin ones would be using HTTP endpoints that would be specified for their respective exporters only. Also, there are file based exporters or Kafka based one, for them the endpoint is not relevant.

Shephalimittal replied:
@reta I have revised the PR to remove support for configurable settings as of now and instead of checking create() , added check for getDefault() method in the exporter (with default settings). Also instead of creating provider, called  OTLPGrpcSpanExporter directly. This would provide users to visualise spans using OTLPGrpcSpanExporter.

For providing support for configurable settings, I will soon raise a RFC for discussion on use of adding auto-configurable sdk or any alternate solutions to read the configurable settings in custom exporters, samplers etc.
Let me know, in case of any concerns?

reta commented on plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/exporter/OtlpGrpcSpanExporterProvider.java:
```diff
@@ -0,0 +1,64 @@
+*/
+public final class OtlpGrpcSpanExporterProvider implements SpanExporter {
+
+    private OtlpGrpcSpanExporter delegate;
```
The `OtlpGrpcSpanExporter` under the hood uses service discovery to find out suitable `GrpcSenderProvider`, it seems like you are inclined to use `OkHttpGrpcSenderProvider` but I don't see this dependency anywhere, what sender provider is being used?

reta replied:
The `OtlpGrpcSpanExporter` under the hood uses service discovery to find out suitable `GrpcSenderProvider`, it seems like you are inclined to use `OkHttpGrpcSenderProvider` but I don't see this dependency anywhere, what sender provider is being used?

reta commented:
@Shephalimittal my sincere apologies for the delay, I was confused how this implementation supposed to work in general (it looked to me like we are missing dependencies) but it is cleared out now: we use pretty old Otel version 1.26.0 whereas the latest release is 1.29.0. 

The `OtlpGrpcSpanExporter` has been refactored to use the service discovery mechanism instead of bundling the OkHttp based span exported by default. I think we should bring our dependencies up to date, I will send over the pull request shortly.

Shephalimittal commented:
> @Shephalimittal my sincere apologies for the delay, I was confused how this implementation supposed to work in general (it looked to me like we are missing dependencies) but it is cleared out now: we use pretty old Otel version 1.26.0 whereas the latest release is 1.29.0.
> 
> The `OtlpGrpcSpanExporter` has been refactored to use the service discovery mechanism instead of bundling the OkHttp based span exported by default. I think we should bring our dependencies up to date, I will send over the pull request shortly.

@reta I have removed the wrapper over OTLPGrpcSpanExporter here and added support to use it with default implementation instead of configuring it through settings. We need some basic support of exporter for testing and visualising spans for adding instrumentations in various code paths. We can incrementally provide support for adding configurable settings in exporter.

reta commented:
> We need some basic support of exporter for testing and visualising spans for adding instrumentations in various code paths. We can incrementally provide support for adding configurable settings in exporter.

@Shephalimittal sure, please rebase, the change needs adjustment for Otel Java Sdk 1.30.1

reta commented on plugins/telemetry-otel/build.gradle:
```diff
@@ -30,6 +30,12 @@ dependencies {
+  api "io.opentelemetry:opentelemetry-exporter-otlp:${versions.opentelemetry}"
+  api "io.opentelemetry:opentelemetry-exporter-common:${versions.opentelemetry}"
+  api "io.opentelemetry:opentelemetry-exporter-otlp-common:${versions.opentelemetry}"
+  api "com.squareup.okhttp3:okhttp:4.10.0"
```
Please use:

```suggestion
    runtimeOnly "io.opentelemetry:opentelemetry-exporter-sender-okhttp:${versions.opentelemetry}"
```

Shephalimittal replied:
Will still need to keep okhttp3 dependency.

Getting this error after removing this api "com.squareup.okhttp3:okhttp:4.10.0"

```
java.lang.IllegalStateException: SpanExporter instantiation failed for class [io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.lambda$instantiateSpanExporter$0(OTelSpanExporterFactory.java:78) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:569) ~[?:?]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.instantiateSpanExporter(OTelSpanExporterFactory.java:58) ~[?:?]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.create(OTelSpanExporterFactory.java:49) ~[?:?]
	at org.opensearch.telemetry.tracing.OTelResourceProvider.lambda$get$0(OTelResourceProvider.java:53) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:318) ~[?:?]
	at org.opensearch.telemetry.tracing.OTelResourceProvider.get(OTelResourceProvider.java:50) ~[?:?]
	at org.opensearch.telemetry.OTelTelemetryPlugin.telemetry(OTelTelemetryPlugin.java:62) ~[?:?]
	at org.opensearch.telemetry.OTelTelemetryPlugin.getTelemetry(OTelTelemetryPlugin.java:53) ~[?:?]
	at org.opensearch.telemetry.TelemetryModule.<init>(TelemetryModule.java:28) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.node.Node.<init>(Node.java:596) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.node.Node.<init>(Node.java:400) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:242) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:404) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:180) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:171) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138) [opensearch-cli-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cli.Command.main(Command.java:101) [opensearch-cli-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:137) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:103) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
Caused by: java.lang.ClassNotFoundException: okhttp3.Interceptor
	at java.net.URLClassLoader.findClass(URLClassLoader.java:445) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:592) ~[?:?]
	at java.net.FactoryURLClassLoader.loadClass(URLClassLoader.java:872) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at io.opentelemetry.exporter.sender.okhttp.internal.OkHttpGrpcSenderProvider.createSender(OkHttpGrpcSenderProvider.java:42) ~[?:?]
	at io.opentelemetry.exporter.internal.grpc.GrpcExporterBuilder.build(GrpcExporterBuilder.java:158) ~[?:?]
	at io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporterBuilder.build(OtlpGrpcSpanExporterBuilder.java:189) ~[?:?]
	at io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter.getDefault(OtlpGrpcSpanExporter.java:33) ~[?:?]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.lambda$instantiateSpanExporter$0(OTelSpanExporterFactory.java:71) ~[?:?]
```

Shephalimittal replied:
Removing api "com.squareup.okhttp3:okhttp:4.10.0" is causing error
```
java.lang.IllegalStateException: SpanExporter instantiation failed for class [io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.lambda$instantiateSpanExporter$0(OTelSpanExporterFactory.java:78) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:569) ~[?:?]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.instantiateSpanExporter(OTelSpanExporterFactory.java:58) ~[?:?]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.create(OTelSpanExporterFactory.java:49) ~[?:?]
	at org.opensearch.telemetry.tracing.OTelResourceProvider.lambda$get$0(OTelResourceProvider.java:53) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:318) ~[?:?]
	at org.opensearch.telemetry.tracing.OTelResourceProvider.get(OTelResourceProvider.java:50) ~[?:?]
	at org.opensearch.telemetry.OTelTelemetryPlugin.telemetry(OTelTelemetryPlugin.java:62) ~[?:?]
	at org.opensearch.telemetry.OTelTelemetryPlugin.getTelemetry(OTelTelemetryPlugin.java:53) ~[?:?]
	at org.opensearch.telemetry.TelemetryModule.<init>(TelemetryModule.java:28) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.node.Node.<init>(Node.java:596) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.node.Node.<init>(Node.java:400) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:242) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:404) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:180) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:171) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138) [opensearch-cli-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cli.Command.main(Command.java:101) [opensearch-cli-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:137) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:103) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
Caused by: java.lang.ClassNotFoundException: okhttp3.Interceptor
	at java.net.URLClassLoader.findClass(URLClassLoader.java:445) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:592) ~[?:?]
	at java.net.FactoryURLClassLoader.loadClass(URLClassLoader.java:872) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at io.opentelemetry.exporter.sender.okhttp.internal.OkHttpGrpcSenderProvider.createSender(OkHttpGrpcSenderProvider.java:42) ~[?:?]
	at io.opentelemetry.exporter.internal.grpc.GrpcExporterBuilder.build(GrpcExporterBuilder.java:158) ~[?:?]
	at io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporterBuilder.build(OtlpGrpcSpanExporterBuilder.java:189) ~[?:?]
	at io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter.getDefault(OtlpGrpcSpanExporter.java:33) ~[?:?]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.lambda$instantiateSpanExporter$0(OTelSpanExporterFactory.java:71) ~[?:?]
```

reta replied:
@Shephalimittal I think we don't even need that, since the baseline is JDK-11, we could use JDK sender with no additional dependencies:

```
api "io.opentelemetry:opentelemetry-exporter-sender-jdk:${versions.opentelemetry}-alpha"
```

Shephalimittal replied:
I think as per https://github.com/open-telemetry/opentelemetry-java/issues/5572, We will need to add opentelemetry-exporter-sender-okhttp or opentelemetry-exporter-sender-grpc-upstream

Getting error:
```
java.lang.IllegalStateException: No GrpcSenderProvider found on classpath. Please add dependency on opentelemetry-exporter-sender-okhttp or opentelemetry-exporter-sender-grpc-upstream
[2023-09-21T01:23:13,989][ERROR][o.o.b.Bootstrap          ] [node-2] Exception
java.lang.IllegalStateException: SpanExporter instantiation failed for class [io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter]
	at org.opensearch.telemetry.tracing.exporter.OTelSpanExporterFactory.lambda$instantiateSpanExporter$0(OTelSpanExporterFactory.java:80) ~[?:?]
```

reta replied:
:o this is surprise, sorry about wrong side track, it seems like we indeed have to add the dependency on okhttp, please use `runtimeOnly` scope as was suggested previously, this dependency should only appear in the plugin bundle.

Shephalimittal replied:
Ack.

reta commented on plugins/telemetry-otel/build.gradle:
```diff
@@ -30,6 +30,13 @@ dependencies {
+  api "io.opentelemetry:opentelemetry-exporter-otlp:${versions.opentelemetry}"
+  api "io.opentelemetry:opentelemetry-exporter-common:${versions.opentelemetry}"
+  api "io.opentelemetry:opentelemetry-exporter-otlp-common:${versions.opentelemetry}"
+  api "com.squareup.okhttp3:okhttp:4.10.0"
```
@Shephalimittal the Okhttp dependencies should also be under `runtimeOnly`:

```
runtimeOnly "org.jetbrains.kotlin:kotlin-stdlib:${versions.kotlin}"
runtimeOnly "com.squareup.okhttp3:okhttp:4.11.0"
runtimeOnly "com.squareup.okio:okio-jvm:3.5.0"
```

reta commented:
Thank you @Shephalimittal !
